lib/cn/csched_sp3.c

```diff
@@ -1134,37 +1151,105 @@ sp3_process_ingest(struct sp3 *sp)
 static void
 sp3_process_dirtylist(struct sp3 *sp)
 {
-    struct cn_tree *tree;
+    struct cn_tree *tree, *tree_next;
 
-    list_for_each_entry(tree, &sp->mon_tlist, ct_sched.sp3t.spt_tlink) {
+    list_for_each_entry_safe(tree, tree_next, &sp->mon_tlist, ct_sched.sp3t.spt_tlink) {
         struct sp3_tree *spt = tree2spt(tree);
-        struct sp3_node *spn;
+        struct sp3_node *spn, *spn_next;
+        struct list_head joined;
+        uint ndirty = 0;
         void *lock;
 
+        if (atomic_read(&sp->sp_dirty_count) == 0)
+            return;
+
+        INIT_LIST_HEAD(&joined);
+
         rmlock_rlock(&tree->ct_lock, &lock);
+        mutex_lock(&spt->spt_dlist_lock);
 
         /* Verify that the action didn't dislodge the root node
          * from the head of the nodes list.
          */
         assert(tree->ct_root == list_first_entry(&tree->ct_nodes, struct cn_tree_node, tn_link));
 
-        mutex_lock(&spt->spt_dlist_lock);
+        list_for_each_entry_safe(spn, spn_next, &spt->spt_dlist, spn_dlink) {
+            struct cn_tree_node *tn = spn2tn(spn);
 
-        while (NULL != (spn = list_first_entry_or_null(&spt->spt_dlist, typeof(*spn), spn_dlink))) {
             if (!spn->spn_initialized)
                 sp3_node_init(sp, spn);
 
-            sp3_dirty_node_locked(sp, spn2tn(spn));
-
-            /* Delete from list and init spn's dlink so we can detect whether or not an spn is on
-             * the list.
+            /* Delete from list and init spn's dlink so we can detect whether
+             * or not an spn is on the list.
              */
             list_del_init(&spn->spn_dlink);
+            ndirty++;
+
+            /* If this leaf node has no route node pointer then it must be the
+             * left node of a successful join, so we put it on the "joined" list
+             * and will remove it from the tree it under the tree write lock.
+             */
+            if (cn_node_isleaf(tn) && !tn->tn_route_node) {
+                list_add_tail(&spn->spn_dlink, &joined);
+                continue;
+            }
+
+            sp3_dirty_node_locked(sp, spn2tn(spn));
         }
 
+        if (ndirty > 0) {
+            atomic_sub(&sp->sp_dirty_count, ndirty);
+            spt->spt_dlist_cnt += ndirty;
+        }
         mutex_unlock(&spt->spt_dlist_lock);
-
         rmlock_runlock(lock);
+
+        /* Move busy trees to the front of the tree list to reduce
+         * the number of trees csched has to visit to find work.
+         */
+        if (spt->spt_dlist_cnt > 16) {
+            list_del(&spt->spt_tlink);
+            list_add(&spt->spt_tlink, &sp->mon_tlist);
+
+            spt->spt_dlist_cnt = 0;
+        }
```

Comment:
Crude, but effective.  Open to suggestions for an inexpensive MRU list of trees...